From e2cac47b33ab8e77936f2f0116a4b2ce8bac3eda Mon Sep 17 00:00:00 2001
From: Clement Faure <clement.faure@nxp.com>
Date: Thu, 16 Jun 2022 10:03:24 +0200
Subject: [PATCH] core: pta: return error code when failing to deserialize
 saved key

When deserializing the key, TEE_SUCCESS code is returned even the bignum
deserialization of one of the key component fails.

Make sure to free the allocated key in case of an error in the
deserialization of one of the key component.

Fixes: 7e05ec25b ("core: pta: add remote attestation PTA")
Signed-off-by: Clement Faure <clement.faure@nxp.com>
Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>
Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>
---
 core/pta/attestation.c | 15 +++++++++------
 1 file changed, 9 insertions(+), 6 deletions(-)

diff --git a/core/pta/attestation.c b/core/pta/attestation.c
index 671fc4d6..4a3f70f5 100644
--- a/core/pta/attestation.c
+++ b/core/pta/attestation.c
@@ -180,23 +180,26 @@ static TEE_Result deserialize_key(uint8_t *buf, size_t buf_sz)

	res = allocate_key();
	if (res)
-		goto out;
+		return res;

	sz = deserialize_bignum(p, buf_sz, key->e);
	if (!sz)
-		goto out;
+		goto err;
	p += sz;
	buf_sz -= sz;
	sz = deserialize_bignum(p, buf_sz, key->d);
	if (!sz)
-		goto out;
+		goto err;
	p += sz;
	buf_sz -= sz;
	sz = deserialize_bignum(p, buf_sz, key->n);
	if (!sz)
-		goto out;
-out:
-	return res;
+		goto err;
+
+	return TEE_SUCCESS;
+err:
+	free_key();
+	return TEE_ERROR_GENERIC;
 }

 static TEE_Result sec_storage_obj_read(TEE_UUID *uuid, uint32_t storage_id,
--
2.25.1
